### PR TITLE
[SPARK-9700] Pick default page size more intelligently.

### PR DESCRIPTION
**Pull request by rxin**
Previously, we use 64MB as the default page size, which was way too big for a lot of Spark applications (especially for single node).

This patch changes it so that the default page size, if unset by the user, is determined by the number of cores available and the total execution memory available.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala [hunk 1]
```diff
@@ -124,7 +130,23 @@ private[spark] class ShuffleMemoryManager(maxMemory: Long) extends Logging {
   }
 }
 
-private object ShuffleMemoryManager {
+object ShuffleMemoryManager {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala]: needs to be `private[something]`

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala [hunk 2]
```diff
@@ -34,11 +37,14 @@ import org.apache.spark.{Logging, SparkException, SparkConf, TaskContext}
...
  * this set changes. This is all done by synchronizing access on "this" to mutate state and using
  * wait() and notifyAll() to signal changes.
+ *
+ * Use `ShuffleMemoryManager.create()` factory method to create a new instance.
  */
-private[spark] class ShuffleMemoryManager(maxMemory: Long) extends Logging {
-  private val taskMemory = new mutable.HashMap[Long, Long]()  // taskAttemptId -> memory bytes
+private[spark]
+class ShuffleMemoryManager protected (val maxMemory: Long, val pageSizeBytes: Long)
+  extends Logging {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleMemoryManager.scala]: I usually do
```
private[spark] class ShuffleMemoryManager ptoected(
    val maxMemory: Long,
    val pageSizeBytes: Long)
  extends Logging {
```
not a big deal...

**Comment by andrewor14** [comment]
looks great!

**Comment by rxin** [comment]
Feedback addressed.

**Comment by andrewor14** [comment]
LGTM. Eventually we should just remove the `pageSizeBytes` from the constructors of all the unsafe things, since we already pass in the `ShuffleMemoryManager` there.

**Comment by rxin** [comment]
Going to merge this.

**Pull request closed without merging**